{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Directions for inputting a VCF (Variant Call Format) SNP file into Haploview\n",
    "\n",
    "#### by [Margaret Antonio](github.com/mmlantonio/) 16.08.24 \n",
    "\n",
    "#### Overview\n",
    "\n",
    "Haploview has useful visualization and analysis tools for haplotype and LD. We can convert a VCF file into PLINK format (.ped and .map files) for input as LINKAGE FORMAT in Haploview. \n",
    "\n",
    "\n",
    "#### TOOLS Used (see linked manuals for more information)\n",
    "\n",
    "* [VCF TOOLS](http://vcftools.sourceforge.net/man_latest.html)\n",
    "* [PLINK](http://pngu.mgh.harvard.edu/~purcell/plink/)\n",
    "* [HAPLOVIEW](https://www.broadinstitute.org/scientific-community/science/programs/medical-and-population-genetics/haploview/haploview)\n",
    "\n",
    "\n",
    "#### Command line\n",
    "\n",
    "1. Remove indels from VCF file (Haploview LINKAGE FORMAT does not accept INDELS)\n",
    "\n",
    "    ```bash\n",
    "    vcftools --vcf ../inputs/snpEff/snpEff_GATK_filtered.vcf --remove-indels --recode --recode-INFO-all --out gatk_filt_SNPonly \n",
    "    ```\n",
    "\n",
    "2. Convert VCF to PLINK format \n",
    "\n",
    "    ```bash\n",
    "    vcftools --vcf gatk_filt_SNPonly.recode.vcf --plink --out gatk_filt_SNPonly\n",
    "    ```\n",
    "\n",
    "3. OPTIONAL. In plink, update family and individual IDs. VCF files only have individual IDs, so this is the time to specify the family IDs. If no Family ID, the just use the Individual ID.)\n",
    "\n",
    "    ```bash\n",
    "   plink --file gatk_filt_SNPonly --update-ids update/newIDs.txt --recode --out edit5_gatk_filt_SNPonly \n",
    "    \n",
    "    #ID file has 4 columns: Old-FamilyID Old-IndivID New-FamilyID New-IndivID\n",
    "    #Example=> SM:DH10 SM:DH10 DH10 DH10\n",
    "    #          SM:DH12 SM:DH12 DH12 DH12\n",
    "    ```\n",
    "\n",
    "4. OPTIONAL. In Plink, update phenotypes where 1=unaffected and 2=affected. \n",
    "\n",
    "    ```bash\n",
    "   plink --file edit5_gatk_filt_SNPonly --pheno update/newPheno.txt --recode --out edit6_gatk_filt_SNPonly \n",
    "    \n",
    "     #Phenotype file has 3 columns: FamilyID IndivID Phenotype\n",
    "     #Example=> DH8     DH8\t    2\n",
    "     #          Ki2007\tKi2007\t1\n",
    "    ```\n",
    "\n",
    "5. OPTIONAL. In Plink, update sex.\n",
    "\n",
    "    ```bash\n",
    "   plink --file edit6_gatk_filt_SNPonly --update-sex update/newSEX.txt --recode --out edit7_gatk_filt_SNPonly \n",
    "    \n",
    "    #File in --update-sex contains 3 columns: FamilyID IndivID Sex. Use 0 for unknown, 1 for female, and 2 for male\n",
    "    ```\n",
    "6. Open .map file as txt file in Excel and remove all other columns, so only 2 columns remain: SNP-marker (chr:pos) and POS(bp).\n",
    "\n",
    "7. Launch Haploview and upload .ped and .map files as DATA FILE and LOCUS INFORMATION, respectively, under the LINKAGE FORMAT data type tab\n",
    "\n",
    "    ```bash\n",
    "    java -jar path/to/Haploview.jar\n",
    "    ```\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
